runtime/src/utils/buffer/write.rs

```diff
@@ -176,153 +210,97 @@ impl<B: Blob> Blob for Write<B> {
         }
 
-        // Case 3: Read spans both blob and buffer
-        let blob_bytes = (buffer_start - offset) as usize;
-        let buffer_bytes = data_len - blob_bytes;
-
-        // Read from blob first
-        let blob_part = vec![0u8; blob_bytes];
-        let blob_part = inner.blob.read_at(blob_part, offset).await?;
-
-        // Copy blob data and buffer data to result
-        buf.as_mut()[..blob_bytes].copy_from_slice(blob_part.as_ref());
-        buf.as_mut()[blob_bytes..].copy_from_slice(&inner.buffer[..buffer_bytes]);
+        // If bytes remain, read directly from the blob.
+        let blob_part = self.blob.read_at(vec![0u8; remaining], offset).await?;
+        buf.as_mut()[..remaining].copy_from_slice(blob_part.as_ref());
 
         Ok(buf)
     }
 
-    async fn write_at(&self, buf: impl Into<StableBuf> + Send, offset: u64) -> Result<(), Error> {
-        // Acquire a write lock on the inner state
-        let mut inner = self.inner.write().await;
+    async fn write_at(&self, data: impl Into<StableBuf> + Send, offset: u64) -> Result<(), Error> {
+        // Prepare `buf` to be written.
+        let data = data.into();
+        let data_len = data.len(); // number of bytes to write
 
-        // Prepare the buf to be written
-        let buf = buf.into();
-        let data = buf.as_ref();
-        let data_len = data.len();
+        // Ensure the write doesn't overflow.
+        let end_offset = offset
+            .checked_add(data_len as u64)
+            .ok_or(Error::OffsetOverflow)?;
 
-        // Current state of the buffer in the blob
-        let buffer_start = inner.position;
-        let buffer_end = buffer_start + inner.buffer.len() as u64;
+        // Acquire a write lock on the buffer.
+        let mut buffer = self.buffer.write().await;
 
-        // Case 1: Simple append to buffered data (most common case)
-        if offset == buffer_end {
-            return inner.write(buf).await;
+        // Case 1: Write falls entirely within the buffer's current range and can be merged.
+        if buffer.merge(data.as_ref(), offset) {
+            return Ok(());
         }
 
-        // Case 2: Write can be merged into existing buffer.
-        //
-        // This handles overwrites and extensions within the buffer's current capacity,
-        // including writes that create gaps (filled with zeros) in the buffer.
-        let can_merge_into_buffer = offset >= buffer_start
-            && (offset - buffer_start) + data_len as u64 <= inner.capacity as u64;
-        if can_merge_into_buffer {
-            let buffer_offset = (offset - buffer_start) as usize;
-            let required_len = buffer_offset + data_len;
-
-            // Expand buffer if necessary (fills with zeros)
-            if required_len > inner.buffer.len() {
-                inner.buffer.resize(required_len, 0);
-            }
+        // Case 2: Write cannot be merged, so flush the buffer if the range overlaps, and check if
```

Comment:
Cases kept here but removed above. I think we should be consistent.

Reply:
You mean from the read_at method? Got rid of that because there are no longer exclusive cases (more or less) -- you just read as much as you can from the buffer (if anything), then if there are more bytes to read, you read them from the blob.  But I guess write is also clear enough without casing (and there is some fall through).  So got rid of the case.

Reply:
Sounds good